(PUP-11123) make Function unwrap more tolerant

cocker-cc opened the pull request:
- as there is no Function `unwrap_if_necessary()` we make the Function `unwrap()`
  more tolerant and let it return a non-sensitive Value just like it is

joshcooper commented:
Hi @cocker-cc, thanks for submitting a pull request! Can you provide more detail about your use case? Is the value originating in hiera or in a different module?

cocker-cc commented:
> Hi @cocker-cc, thanks for submitting a pull request! Can you provide more detail about your use case? Is the value originating in hiera or in a different module?

In the Process of transitioning several Component-Modules to be sensitive-aware, but stay backwards-compatible, I frequently have a Block like:
```
class blubb(
…
  Variant[String, [Sensitive[String]] $password,
…
) {
  $password_unsensitive = if $password =~ Sensitive {
    $password.unwrap
  } else {
    $password
  }
}

```

With the suggested Pullrequest this would reduce to just
```
  $password_unsensitive = $password.unwrap
```

ekohl commented on lib/puppet/functions/unwrap.rb:
```diff
@@ -29,12 +30,12 @@
 Puppet::Functions.create_function(:unwrap) do
   dispatch :unwrap do
-    param 'Sensitive', :arg
+    param 'Any', :arg
```
Perhaps it would be cleaner to have 2 dispatch methods? So it would become
```ruby
dispatch :unwrap_sensitive do
  param 'Sensitive', :arg
  optional_block_param
end

dispatch :unwrap_other do
  param 'Any', :arg
  optional_block_param
end
```

cocker-cc replied:
> Perhaps it would be cleaner to have 2 dispatch methods?   

Suggestion implemented.

cocker-cc replied:
I just wonder, if `dispatch` is deterministic here by only the Order of the two Declarations.  Because `Any` would also match `Sensitive`…

ekohl replied:
Yes, that's something I assumed but https://puppet.com/docs/puppet/7/functions_ruby_signatures.html isn't very explicit in it.

joshcooper replied:
According to [4.x function specification](https://github.com/puppetlabs/puppet-specifications/blob/8db1ac65aa6cc6d84bbe0c366ee4b9378d9bfcbd/language/func-api.md#defining-a-typed-dispatch) the behavior is sensitive to the order:

> the signatures are tested in the order they appear in the definition of the function, thus, if a very generic entry is placed first it will always win.

cocker-cc replied:
👍

ekohl replied:
I don't have permission to mark this as resolved, but this is implemented now.

ekohl reviewed (commented):
I also thought about this. It would certainly be useful to be able to treat `Sensitive[String]` and `String` the same in code. Right now I consider `Sensitive` mostly useless and just a painful thing which is why I've never bothered to use it. Being able to say "I need the raw value now, I know what I'm doing" vs "just log what you have and if it's sensitive, obscure the value" can certainly be a step in the right direction.

joshcooper commented:
Thanks for your contribution @cocker-cc! Could you file a JIRA ticket in the PUP project and `git commit --amend` your commit summary with `(PUP-XXX) ...` This is necessary for our release automation.

Also I wanted to note that [`Sensitive.new` accepts either `Sensitive` or `Any`](https://github.com/puppetlabs/puppet/blob/7.8.0/lib/puppet/pops/types/p_sensitive_type.rb#L54-L70), so it makes sense that its inverse (`unwrap`) does the same.

cocker-cc commented:
> Could you file a JIRA ticket in the PUP project and `git commit --amend` your commit summary with `(PUP-XXX) ...` This is necessary for our release automation.

Done.

> Also I wanted to note that [`Sensitive.new` accepts either `Sensitive` or `Any`](https://github.com/puppetlabs/puppet/blob/7.8.0/lib/puppet/pops/types/p_sensitive_type.rb#L54-L70), so it makes sense that its inverse (`unwrap`) does the same.

Great.  I copied the Variablenames from there to keep it consistent.

joshcooper reviewed (approved):
LGTM. Note `main` branch is for 7.x releases. I'd be fine releasing this in 6.x, you'd just need to retarget your PR against the `6.x` branch.

cocker-cc commented:
> Note `main` branch is for 7.x releases. I'd be fine releasing this in 6.x, you'd just need to retarget your PR against the `6.x` branch.

I rebased it against `6.x` and changed the Merge-Target.

GabrielNagy commented on spec/unit/functions/unwrap_spec.rb:
```diff
@@ -15,6 +15,14 @@
+  it 'just returns a non-sensitive value' do
+    code = <<-CODE
+      $non_sensitive = "12345"
+      notice("value is still ${non_sensitive}")
```
Shouldn't this test call `unwrap` on the variable?

cocker-cc replied:
👍 Yes, you're absolutely right!  I corrected the Line to:
```
notice("value is still ${non_sensitive.unwrap}")
```

ekohl commented:
This was merged to `6.x`, but it should also be in `main`. In fact, normally I'd say that it should first be merged to `main` and then cherry picked to `6.x`.

GabrielNagy commented:
@ekohl we have automation that promotes/merges up commits from `6.x` to `main` once our internal nightly acceptance CI passes

joshcooper commented:
Thanks for your contribution @cocker-cc! This should be released in the next set of releases in July.
